### ISSUE TITLE
Does ClickHouse now support Redis cluster as a dictionary source?

### ISSUE THREAD

**Issue by fsgdgsdfgsgd** [ClickHouse/ClickHouse#27123]
I read the document on using Redis as a dictionary source: https://clickhouse.tech/docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources/#dicts-external_dicts_dict_sources-redis
But I didn't find instructions on cluster connection. 
Is there a effective way to connect  to the Redis cluster?

**Comment by filimonov** [comment]
AFAIR you can talk whole redis cluster via single node / port and standard protocol (just like with a single node), isn't it? 

If you need to pass several IP of different Redis servers to have HA - it's not supported, but you can use loadbalancer in front of Redis cluster which will route the ClickHouse to one of alive servers.

**Comment by lz19970205** [comment]
> AFAIR you can talk whole redis cluster via single node / port and standard protocol (just like with a single node), isn't it?
Not exactly. I noticed that when I use redis-cli to connect to the whole cluster via single node, if I do not declare the -c option, this node will not automatically forward my query to the right node, but only return an error message.
![image](https://user-images.githubusercontent.com/29115270/127988706-0d6c9803-4688-4c03-a4fc-0fe9fb617f28.png)

I also tested it in CH. When I try to get a value that stored on other nodes, it returns a null value.
![image](https://user-images.githubusercontent.com/29115270/127989775-3c1c9901-59ec-4ec4-8fa7-08babfa9cb33.png)

I think it might be better to add an option to redis configuration. This option indicates whether to connect as a cluster.

**Comment by filimonov** [comment]
https://serverfault.com/questions/812156/redis-cluster-error-moved 

It seems like Poco::Redis don't support those redirects...

**Comment by alexey-milovidov** [comment]
Answering the original question definitively: **no**, Redis Cluster is not supported as a dictionary source, and the reason is the client library. `RedisDictionarySource` / `RedisSource` are built on `Poco::Redis` (`src/Dictionaries/RedisSource.cpp` issues plain `MGET`/`HMGET` commands), which has no notion of the cluster protocol — it does not read the `MOVED`/`ASK` redirections, does not fetch the slot map, and does not open connections to the other nodes. That is why you see a `NULL` for keys that live on another node instead of an error: the redirection reply is not recognised as a hit.

Supporting it properly means a cluster-aware client (slot map, `MOVED`/`ASK` handling, per-node connection pool, resharding invalidation), i.e. replacing `Poco::Redis`, not adding a configuration flag.

The practical workaround is unchanged: put something cluster-aware in front of ClickHouse that presents a single-node view, and point the dictionary at it.

Closing as answered / not planned. If cluster support becomes a real need, a fresh issue framed as "replace Poco::Redis with a cluster-capable client" (affecting the `Redis` dictionary source *and* the `Redis` table engine) is the right shape for it.
